### PR TITLE
Sphinx apidoc does not process PEP-0420 implicit namespaces

### LINKED ISSUE #2949: Sphinx apidoc does not process PEP-0420 implicit namespaces
Under [PEP-0420](https://www.python.org/dev/peps/pep-0420/) implicit namespaces are possible. 

That is a path structure like this:

```
a/b/c/d/__init__.py
a/b/c/d/e/__init__.py
```

produces two modules whose qualified names are "a.b.c.d" and "a.b.c.d.e".

Assuming `a/b/c` is located in `src/main/python` the invoking `sphinx-apidoc -o docs src/main/python` does not see and does not process modules `a.b.c.d` and `a.b.c.d.e`. However, if `sphinx-apidoc -o docs src/main/python/a/b/c/` is invoked then modules `d` and `e` are observed and RSTs are generated, albeit with incorrect qualified names due to wrong package path.

The compliant behavior is to follow namespace resolution in PEP-0420 and traverse package paths to arbitrary depth treating directories without `__init__.py` in the manner consistent with PEP-0420.

### PR DESCRIPTION
**Pull request by arcivanov**
fixes #2949, connected to #2949

### CODE REVIEW

**Comment by arcivanov** [comment]
@tk0miya could you also weigh-in here?

**Comment by tk0miya** [comment]
I'll take a look later :-) Please wait a bit

File: sphinx/apidoc.py [hunk 1]
```diff
@@ -295,6 +306,10 @@ def main(argv=sys.argv):
                       dest='modulefirst',
                       help='Put module documentation before submodule '
                       'documentation')
+    parser.add_option('--pep-420', action='store_true',
```
Comment (tk0miya) [inline sphinx/apidoc.py]: I prefer to more explicit naming :-)
How about `--implicit-namespace`?

Comment (tk0miya) [inline sphinx/apidoc.py]: Could you update `doc/invocation.rst` about this option?

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits

**Comment by tk0miya** [comment]
In addition, please add CHANGES entry.

File: sphinx/apidoc.py (hunk 1 above)
Comment (arcivanov) [reply sphinx/apidoc.py]: Rgr that.

Comment (arcivanov) [reply sphinx/apidoc.py]: Done.

Comment (arcivanov) [reply sphinx/apidoc.py]: Done.

**Comment by arcivanov** [comment]
@tk0miya  Ready for final review and commit.

File: doc/invocation.rst [hunk 2]
```diff
@@ -453,6 +453,15 @@ The :program:`sphinx-apidoc` script has several options:
    to default values, but you can influence the most important ones using the
    following options.
 
+.. option:: --implicit-namespaces
+
+   By default `sphinx-apidoc` processes sys.path searching for modules only.
+   Python 3.3 introduced PEP-0420 implicit namespaces that allow module path
```
Comment (tk0miya) [inline doc/invocation.rst]: It would be nice if using `:pep:` role. hyperlinks will help readers.

**Review by tk0miya — APPROVED** [review]
LGTM!

File: doc/invocation.rst (hunk 2 above)
Comment (arcivanov) [reply doc/invocation.rst]: One sec, will fix.

**Comment by tk0miya** [comment]
@shimizukawa do you have any comments?
If not, I will merge this soon (maybe in this weekend).

File: doc/invocation.rst (hunk 2 above)
Comment (arcivanov) [reply doc/invocation.rst]: done

Comment (tk0miya) [reply doc/invocation.rst]: Thanks!

**Comment by tk0miya** [comment]
Merged.
Thank you for contribution!

**Pull request merged**